### PR TITLE
http: add client identity

### PR DESCRIPTION
**Pull request by binarytrails**
- client test sample: [https_2waystls.cpp](https://github.com/binarytrails/various/blob/master/cpp/opendht/https_2waystls.cpp#L48
)
- server dhtproxy feature: #432

### CODE REVIEW

File: include/opendht/dht_proxy_client.h [hunk 1]
```diff
@@ -328,6 +328,7 @@ class OPENDHT_PUBLIC DhtProxyClient final : public DhtInterface {
     void cancelAllOperations();
 
     std::shared_ptr<dht::crypto::Certificate> serverCertificate_;
+    std::shared_ptr<dht::crypto::Certificate> clientCertificate_;
```
Comment (aberaud) [inline include/opendht/dht_proxy_client.h]: Identity

File: include/opendht/dhtrunner.h [hunk 2]
```diff
@@ -63,7 +63,8 @@ class OPENDHT_PUBLIC DhtRunner {
         std::string push_token {};
         bool peer_discovery {false};
         bool peer_publish {false};
-        std::shared_ptr<dht::crypto::Certificate> client_cert;
+        std::shared_ptr<dht::crypto::Certificate> client_ca;
```
Comment (aberaud) [inline include/opendht/dhtrunner.h]: server_ca?

File: include/opendht/http.h [hunk 3]
```diff
@@ -85,17 +86,15 @@ class OPENDHT_PUBLIC Connection
 {
 public:
     Connection(asio::io_context& ctx, const bool ssl = true, std::shared_ptr<dht::Logger> l = {});
-    Connection(asio::io_context& ctx, std::shared_ptr<dht::crypto::Certificate> certificate,
-               std::shared_ptr<dht::Logger> l = {});
+    Connection(asio::io_context& ctx, std::shared_ptr<dht::crypto::Certificate> ca_cert,
+               dht::crypto::Identity& identity, std::shared_ptr<dht::Logger> l = {});
```
Comment (aberaud) [inline include/opendht/http.h]: `const dht::crypto::Identity&` or `dht::crypto::Identity`

File: include/opendht/http.h [hunk 4]
```diff
@@ -120,7 +119,9 @@ class OPENDHT_PUBLIC Connection
     std::unique_ptr<socket_t> socket_;
     std::shared_ptr<asio::ssl::context> ssl_ctx_;
     std::unique_ptr<ssl_socket_t> ssl_socket_;
-    std::unique_ptr<asio::const_buffer> certificate_;
+    std::unique_ptr<asio::const_buffer> ca_cert_;
+    std::unique_ptr<asio::const_buffer> client_pk_;
```
Comment (aberaud) [inline include/opendht/http.h]: `pk` is usually used for "public key". For private key simply use `key`

File: src/http.cpp [hunk 5]
```diff
@@ -99,23 +99,40 @@ Connection::Connection(asio::io_context& ctx, const bool ssl, std::shared_ptr<dh
...
-                                                       (std::size_t) cert.size());
-    ssl_ctx_->use_certificate(*certificate_, asio::ssl::context::file_format::pem, ec);
-    if (ec)
-        throw std::runtime_error("Error setting certificate: " + ec.message());
-    else if (logger_)
-        logger_->d("[http:client]  [connection:%i] start https session with %s", id_, certificate->getUID().c_str());
-
+    if (server_ca){
+        auto ca = server_ca->toString(false/*chain*/);
+        server_ca_ = std::make_unique<asio::const_buffer>(static_cast<const void*>(ca.data()),
```
Comment (aberaud) [inline src/http.cpp]: server_ca_ not relevant, inline asio::const_buffer

**Comment by binarytrails** [comment]
testing in progress with https://github.com/savoirfairelinux/opendht/pull/432

**Pull request merged**